src/core/meshcop/dtls.cpp

```diff
@@ -287,7 +290,7 @@ otError Dtls::Setup(bool aClient)
     mbedtls_ssl_conf_handshake_timeout(&mConf, 8000, 60000);
     mbedtls_ssl_conf_dbg(&mConf, HandleMbedtlsDebug, this);
 
-#ifdef MBEDTLS_SSL_SRV_C
+#if defined(MBEDTLS_SSL_SRV_C) && defined(MBEDTLS_SSL_COOKIE_C)
     if (!aClient)
     {
         mbedtls_ssl_cookie_init(&mCookieCtx);
```

Comment:
We will still get hard fault if we failed before reaching here. Because we will call `FreeMbedtls` before  exiting if there are errors. see https://github.com/openthread/openthread/blob/master/src/core/meshcop/dtls.cpp#L343

Suggestion: init cookie at the beginning of this function.

Reply:
@wgtdkp - Perhaps I'm missing something but I believe the changes @mnp222 made to FreeMbedtls will prevent the hard fault you mention.

Reply:
@romacdon cookie will be freed before getting initialized if [this line](https://github.com/openthread/openthread/blob/master/src/core/meshcop/dtls.cpp#L263) fails.
We should do all `mbedtls_*_init(...)` stuff at the beginning of this function.

Reply:
@wgtdkp I see.  @mnp222 made fix to address an issue associated with client side cookies.  Your additional request is to fix a potential issue with the server side cookie.

Reply:
Since the bug is freeing cookie before initialization, I'd like to fix it (both client and server side) in this PR.

Reply:
@wgtdkp yes, understood and I agree.  Thank you.

Reply:
Hi Kang. You are correct and I agree we need to make this safe in all cases.  I just committed a change that only calls free when there is something to be freed, so it should take care of all cases, including the server and including when a failure happens before allocation.

Reply:
@mnp222 Thanks for you update.
```c++
if (mCookieCtx.hmac_ctx.hmac_ctx != NULL)
{
    mbedtls_ssl_cookie_free(&mCookieCtx);
}
```
But I don't think it is a good idea to reference to the implementation of `mbedtls_ssl_cookie_ctx` and depend on it. We should stick to the mbedtls cookie APIs (`mbedtls_ssl_cookie_*`) and they should be enough.

Besides, OpenThread coding style puts declarations and initialization at the beginning of a function, we should do so with cookie.

Reply:
I agree Kang.  I spent more time investigating the reason for the fault, and I discovered the real issue in the documentation for mbedtls md.h:

```
/**                                                                                                                                                                  
 * \brief           This function clears the internal structure of \p ctx and                                                                                        
 *                  frees any embedded internal structure, but does not free                                                                                         
 *                  \p ctx itself.                                                                                                                                   
 *                                                                                                                                                                   
 *                  If you have called mbedtls_md_setup() on \p ctx, you must                                                                                        
 *                  call mbedtls_md_free() when you are no longer using the                                                                                          
 *                  context.                                                                                                                                         
 *                  Calling this function if you have previously                                                                                                     
 *                  called mbedtls_md_init() and nothing else is optional.                                                                                           
 *                  You must not call this function if you have not called                                                                                           
 *                  mbedtls_md_init().                                                                                                                               
 */
void mbedtls_md_free( mbedtls_md_context_t *ctx );
```

So the issue happens when mbedtls_ssl_cookie_free() (which calls mbedtls_md_free()) is called without calling mbedtls_ssl_cookie_init() (which calls mbedtsl_md_init()). 

I fixed the dtls.cpp code so that mbedtls_ssl_cookie_init() is called at the beginning of Setup(), before any failure can occur, and for either client or server.  That way in FreeMbedtls(), mbedtls_ssl_cookie_free() can be called to cleanup regardless of what happened. This fixes the fault for me.

Unfortunately some Travis tests are failing, so I'll have to look into that.

Reply:
@mnp222 it seems the travis failures are all related to commissioning, which is the only feature depends on dtls. I think it is cause by the issue here https://github.com/openthread/openthread/pull/4564#discussion_r384184881